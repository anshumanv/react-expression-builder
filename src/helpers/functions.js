export const functions = [
	{
		value: 'SPLIT (dim, delimiter, occurrence_number)',
		label: 'SPLIT',
		type: 'function',
		key: 'split',
		params: ['dim', 'delimiter', 'occurrence_number'],
		helper:
			'Returns the nth substring divided by a specified delimiter. Index, n, starts from 0'
	},
	{
		value: 'CONCAT (dim1, dim2)',
		label: 'CONCAT',
		key: 'concat',
		params: ['dim1', 'dim2'],
		type: 'function',
		helper: 'Returns the concatenation of two strings.'
	},
	{
		value: 'SUB (dim, starting_at, ending_at)',
		label: 'SUB',
		key: 'sub',
		params: ['dim', 'starting_at', 'ending_at'],
		type: 'function',
		helper:
			'Returns a substring between specified character indices. Index starts from 0'
	},
	{
		value: 'EXTRACT (dim, prefix_string, suffix_string)',
		label: 'EXTRACT',
		key: 'extract',
		params: ['dim1', 'prefix_string', 'suffix_string'],
		type: 'function',
		helper:
			'Returns a substring between the first prefix_string and first suffix_string'
	}
]

export const staticValues = [
	{
		value: 'ACCOUNT',
		label: 'Account',
		type: 'dimension',
		key: 'account',
		helper: 'Account dimension'
	},
	{
		value: 'AD',
		label: 'Ad',
		type: 'dimension',
		key: 'ad',
		helper: 'Account ad'
	}
]
