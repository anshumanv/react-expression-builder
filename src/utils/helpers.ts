export const functions = [
	{
		value: 'SPLIT (dim, delimiter, occurrence_number)',
		label: 'SPLIT',
		type: 'function',
		key: 'split',
		keyLabel: 'f(x)',
		params: ['dim', 'delimiter', 'occurrence_number'],
		helper:
			'Returns the nth substring divided by a specified delimiter. Index, n, starts from 0'
	},
	{
		value: 'CONCAT (dim1, dim2)',
		label: 'CONCAT',
		key: 'concat',
		params: ['dim1', 'dim2'],
		type: 'function',
		keyLabel: 'f(x)',
		helper: 'Returns the concatenation of two strings.'
	},
	{
		value: 'SUB (dim, starting_at, ending_at)',
		label: 'SUB',
		key: 'sub',
		params: ['dim', 'starting_at', 'ending_at'],
		type: 'function',
		keyLabel: 'f(x)',
		helper:
			'Returns a substring between specified character indices. Index starts from 0'
	},
	{
		value: 'EXTRACT (dim, prefix_string, suffix_string)',
		label: 'EXTRACT',
		key: 'extract',
		params: ['dim1', 'prefix_string', 'suffix_string'],
		type: 'function',
		keyLabel: 'f(x)',
		helper:
			'Returns a substring between the first prefix_string and first suffix_string'
	}
]

export const staticValues = [
	{
		value: 'ACCOUNT',
		label: 'Account',
		type: 'dimension',
		keyLabel: 'dim',
		key: 'account',
		helper: 'Account dimension'
	},
	{
		value: 'AD',
		label: 'Ad',
		type: 'dimension',
		keyLabel: 'dim',
		key: 'ad',
		helper: 'Account ad'
	}
]
