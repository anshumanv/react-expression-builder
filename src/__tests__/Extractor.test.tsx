import React from 'react'
import Extractor from '../'
import { cleanup, fireEvent, render } from '../../../common/TestUtil'
import { functions, staticValues } from '../helpers'

afterEach(cleanup)

// functions for the extractor
const options = [...functions, ...staticValues]

const concatFunction = functions[1]
const adDimension = staticValues[1]

const onUpdate = editorState => {
	return editorState.buildExpression()
}

const extractorProps = {
	onChangeFn: onUpdate,
	options: options
}

describe('Extractor renders properly', () => {
	it('Renders correctly', () => {
		const { asFragment } = render(<Extractor {...extractorProps} />)

		// Get the options list which should be shown by default
		const optionsList = document.querySelector('[data-type="expression-list"]')
		expect(optionsList).toBeInTheDocument()

		// Check initial focus on render
		const { activeElement } = document
		expect(activeElement!.parentElement).toHaveAttribute(
			'data-type',
			'expression-input-root'
		)

		// Input field should how focus initially
		expect(asFragment()).toMatchSnapshot()
	})
	it('Scaffolds an expression on typing one', () => {
		const { asFragment } = render(<Extractor {...extractorProps} />)
		// Get the root input element
		const { activeElement } = document
		fireEvent.input(activeElement as any, {
			target: { value: functions[0].key }
		})

		// Check that all params are scaffolded properly
		const expressionInputRoots = document.querySelectorAll(
			'[data-type="expression-input-root"]'
		)
		expect(expressionInputRoots).toHaveLength(3)

		// Check that the expression root is created
		const expressionRoots = document.querySelectorAll(
			'[data-type="expression-root"]'
		)
		expect(expressionRoots).toHaveLength(1)

		expect(asFragment()).toMatchSnapshot()
	})
	it('Hides options list on blur', () => {
		const { asFragment } = render(<Extractor {...extractorProps} />)
		// Initially the input has focus
		// Get the options list which should be shown by default
		const optionsList = document.querySelector('[data-type="expression-list"]')
		expect(optionsList).toBeInTheDocument()

		// create a new element and transfer focus
		const newInput = document.createElement('input')
		newInput.focus()

		expect(optionsList).not.toBeInTheDocument()

		expect(asFragment()).toMatchSnapshot()
	})
	it('Applies correct data attributes', () => {
		const { asFragment } = render(<Extractor {...extractorProps} />)

		// Get the root input element
		const { activeElement } = document
		fireEvent.input(activeElement as any, {
			target: { value: staticValues[1].label }
		})

		// Check for attributes
		expect(activeElement).toHaveAttribute('data-value-type', 'dimension')
		expect(activeElement).toHaveStyle('background-color: #fdedce')

		expect(asFragment()).toMatchSnapshot()
	})
})

describe('Extractor builds data structure properly', () => {
	it('Fires the onChange function', () => {
		const onChangeFn = jest.fn()
		const { asFragment } = render(
			<Extractor {...extractorProps} onChangeFn={onChangeFn} />
		)

		// Get the root input element
		const { activeElement } = document
		fireEvent.change(activeElement as any, {
			target: { value: adDimension.label }
		})

		expect(onChangeFn).toHaveBeenCalled()

		expect(asFragment()).toMatchSnapshot()
	})

	it('Builds correct string on input', () => {
		const onChangeFn = jest.fn(editorState => editorState.buildExpression())
		const { asFragment } = render(
			<Extractor {...extractorProps} onChangeFn={onChangeFn} />
		)

		// Scaffold and expression
		const { activeElement } = document
		fireEvent.change(activeElement as any, {
			target: { value: concatFunction.label }
		})

		// Get scaffolded params
		const expressionParamsInput = document.querySelectorAll(
			'[data-type="expression-input-root"] input'
		)
		expect(expressionParamsInput).toHaveLength(2)
		fireEvent.change(expressionParamsInput[0], {
			target: { value: adDimension.label }
		})
		fireEvent.change(expressionParamsInput[1], { target: { value: '"_"' } })

		expect(onChangeFn).toHaveBeenCalledTimes(3)
		expect(onChangeFn.mock.results[2].value).toEqual('CONCAT (Ad, "_")')

		expect(asFragment()).toMatchSnapshot()
	})
})

describe('Keyboard events work properly', () => {
	it('Should navigate to the expression root on pressing left on first param', () => {
		const { asFragment } = render(<Extractor {...extractorProps} />)

		// Get the root input element
		let { activeElement } = document
		fireEvent.change(activeElement as any, {
			target: { value: concatFunction.key }
		})

		// Active element will be changed to the first param of the expression
		activeElement = document.activeElement

		// Fire left arrow key
		fireEvent.keyDown(activeElement as any, { keyCode: 37 })

		// Check that the focus has shifted to the root of the expression
		activeElement = document.activeElement
		expect(activeElement).toHaveAttribute('data-type', 'expression-root')

		expect(asFragment()).toMatchSnapshot()
	})
	it('Should navigate to the next param on pressing right ', () => {
		const { asFragment } = render(<Extractor {...extractorProps} />)

		// Get the root input element
		let { activeElement } = document
		fireEvent.change(activeElement as any, {
			target: { value: concatFunction.key }
		})

		// Active element will be changed to the first param of the expression
		activeElement = document.activeElement

		// Fire left arrow key
		fireEvent.keyDown(activeElement as any, { keyCode: 39 })

		// Check that the focus has shifted to the root of the expression
		activeElement = document.activeElement
		expect(activeElement!.parentElement).toHaveAttribute(
			'data-type',
			'expression-input-root'
		)
		const allInputs = document.querySelectorAll(
			'[data-type="expression-input-root"]'
		)
		const lastInput = allInputs[1].firstElementChild
		expect(lastInput).toEqual(activeElement)

		// Check that on pressing right, the focus stays for last param
		fireEvent.keyDown(activeElement as any, { keyCode: 39 })
		expect(document.activeElement).toEqual(lastInput)

		expect(asFragment()).toMatchSnapshot()
	})
	it('Should delete the expression root on pressing backspace ', () => {
		const { asFragment } = render(<Extractor {...extractorProps} />)

		// Get the root input element
		let { activeElement } = document
		fireEvent.change(activeElement as any, {
			target: { value: concatFunction.key }
		})

		// Active element will be changed to the first param of the expression
		activeElement = document.activeElement

		// Fire left arrow key
		fireEvent.keyDown(activeElement as any, { keyCode: 37 })

		// Get the expression root
		activeElement = document.activeElement

		// Press backspace
		fireEvent.keyDown(activeElement as any, { keyCode: 8 })
		// Expression root should have been removed by now
		expect(activeElement).not.toBeInTheDocument()

		expect(asFragment()).toMatchSnapshot()
	})
})
