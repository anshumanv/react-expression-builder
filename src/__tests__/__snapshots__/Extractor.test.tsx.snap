// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extractor builds data structure properly Builds correct string on input 1`] = `
<DocumentFragment>
  <span
    class=""
    data-type="expression-root"
    style="display: flex;"
    tabindex="0"
  >
    CONCAT(
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="downshift-20-menu"
      class=""
      data-type="expression-input-root"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="downshift-20-menu"
        aria-labelledby="downshift-20-label"
        autocomplete="off"
        data-valid="true"
        data-value-type="dimension"
        id="downshift-19-input"
        placeholder="dim1"
        style="width: 8px;"
        value="Ad"
      />
      <div
        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
      >
        Ad
      </div>
    </div>
    <div
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-owns="downshift-22-menu"
      class=""
      data-type="expression-input-root"
      role="combobox"
    >
      <input
        aria-activedescendant="downshift-22-item-0"
        aria-autocomplete="list"
        aria-controls="downshift-22-menu"
        aria-labelledby="downshift-22-label"
        autocomplete="off"
        data-valid="true"
        data-value-type="string"
        id="downshift-21-input"
        placeholder="dim2"
        style="width: 8px;"
        value="\\"_\\""
      />
      <div
        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
      >
        "_"
      </div>
    </div>
    )
  </span>
</DocumentFragment>
`;

exports[`Extractor builds data structure properly Fires the onChange function 1`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-owns="downshift-16-menu"
    data-type="expression-input-root"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="downshift-16-menu"
      aria-labelledby="downshift-16-label"
      autocomplete="off"
      data-valid="true"
      data-value-type="dimension"
      id="downshift-15-input"
      style="width: 8px;"
      value="Ad"
    />
    <div
      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
    >
      Ad
    </div>
  </div>
</DocumentFragment>
`;

exports[`Extractor renders properly Applies correct data attributes 1`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-owns="downshift-14-menu"
    data-type="expression-input-root"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="downshift-14-menu"
      aria-labelledby="downshift-14-label"
      autocomplete="off"
      data-valid="true"
      data-value-type="dimension"
      id="downshift-13-input"
      style="width: 8px;"
      value="Ad"
    />
    <div
      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
    >
      Ad
    </div>
  </div>
</DocumentFragment>
`;

exports[`Extractor renders properly Hides options list on blur 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-owns="downshift-12-menu"
    data-type="expression-input-root"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="downshift-12-menu"
      aria-labelledby="downshift-12-label"
      autocomplete="off"
      data-valid="true"
      data-value-type="string"
      id="downshift-11-input"
      style="width: 8px;"
      value=""
    />
    <div
      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
    />
  </div>
</DocumentFragment>
`;

exports[`Extractor renders properly Renders correctly 1`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-owns="downshift-2-menu"
    data-type="expression-input-root"
    role="combobox"
  >
    <input
      aria-activedescendant="downshift-2-item-0"
      aria-autocomplete="list"
      aria-controls="downshift-2-menu"
      aria-labelledby="downshift-2-label"
      autocomplete="off"
      data-valid="true"
      data-value-type="string"
      id="downshift-1-input"
      style="width: 8px;"
      value=""
    />
    <div
      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
    />
    <ul
      aria-labelledby="downshift-2-label"
      data-type="expression-list"
      id="downshift-2-menu"
      role="listbox"
      style="background-color: white; font-weight: normal; position: absolute;"
    >
      <li
        aria-selected="true"
        data-item-type="function"
        data-type="expression-list-item"
        id="downshift-2-item-0"
        role="option"
        style="background-color: rgb(242, 242, 242);"
      >
        <span>
          SPLIT
        </span>
      </li>
      <li
        aria-selected="false"
        data-item-type="function"
        data-type="expression-list-item"
        id="downshift-2-item-1"
        role="option"
      >
        <span>
          CONCAT
        </span>
      </li>
      <li
        aria-selected="false"
        data-item-type="function"
        data-type="expression-list-item"
        id="downshift-2-item-2"
        role="option"
      >
        <span>
          SUB
        </span>
      </li>
      <li
        aria-selected="false"
        data-item-type="function"
        data-type="expression-list-item"
        id="downshift-2-item-3"
        role="option"
      >
        <span>
          EXTRACT
        </span>
      </li>
      <li
        aria-selected="false"
        data-item-type="dimension"
        data-type="expression-list-item"
        id="downshift-2-item-4"
        role="option"
      >
        <span>
          Account
        </span>
      </li>
      <li
        aria-selected="false"
        data-item-type="dimension"
        data-type="expression-list-item"
        id="downshift-2-item-5"
        role="option"
      >
        <span>
          Ad
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Extractor renders properly Scaffolds an expression on typing one 1`] = `
<DocumentFragment>
  <span
    class=""
    data-type="expression-root"
    style="display: flex;"
    tabindex="0"
  >
    SPLIT(
    <div
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-owns="downshift-6-menu"
      class=""
      data-type="expression-input-root"
      role="combobox"
    >
      <input
        aria-activedescendant="downshift-6-item-0"
        aria-autocomplete="list"
        aria-controls="downshift-6-menu"
        aria-labelledby="downshift-6-label"
        autocomplete="off"
        data-valid="true"
        data-value-type="string"
        id="downshift-5-input"
        placeholder="dim"
        style="width: 8px;"
        value=""
      />
      <div
        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
      >
        dim
      </div>
      <ul
        aria-labelledby="downshift-6-label"
        data-type="expression-list"
        id="downshift-6-menu"
        role="listbox"
        style="background-color: white; font-weight: normal; position: absolute;"
      >
        <li
          aria-selected="true"
          data-item-type="function"
          data-type="expression-list-item"
          id="downshift-6-item-0"
          role="option"
          style="background-color: rgb(242, 242, 242);"
        >
          <span>
            SPLIT
          </span>
        </li>
        <li
          aria-selected="false"
          data-item-type="function"
          data-type="expression-list-item"
          id="downshift-6-item-1"
          role="option"
        >
          <span>
            CONCAT
          </span>
        </li>
        <li
          aria-selected="false"
          data-item-type="function"
          data-type="expression-list-item"
          id="downshift-6-item-2"
          role="option"
        >
          <span>
            SUB
          </span>
        </li>
        <li
          aria-selected="false"
          data-item-type="function"
          data-type="expression-list-item"
          id="downshift-6-item-3"
          role="option"
        >
          <span>
            EXTRACT
          </span>
        </li>
        <li
          aria-selected="false"
          data-item-type="dimension"
          data-type="expression-list-item"
          id="downshift-6-item-4"
          role="option"
        >
          <span>
            Account
          </span>
        </li>
        <li
          aria-selected="false"
          data-item-type="dimension"
          data-type="expression-list-item"
          id="downshift-6-item-5"
          role="option"
        >
          <span>
            Ad
          </span>
        </li>
      </ul>
    </div>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="downshift-8-menu"
      class=""
      data-type="expression-input-root"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="downshift-8-menu"
        aria-labelledby="downshift-8-label"
        autocomplete="off"
        data-valid="true"
        data-value-type="string"
        id="downshift-7-input"
        placeholder="delimiter"
        style="width: 8px;"
        value=""
      />
      <div
        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
      >
        delimiter
      </div>
    </div>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="downshift-10-menu"
      class=""
      data-type="expression-input-root"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="downshift-10-menu"
        aria-labelledby="downshift-10-label"
        autocomplete="off"
        data-valid="true"
        data-value-type="string"
        id="downshift-9-input"
        placeholder="occurrence_number"
        style="width: 8px;"
        value=""
      />
      <div
        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
      >
        occurrence_number
      </div>
    </div>
    )
  </span>
</DocumentFragment>
`;

exports[`Keyboard events work properly Should delete the expression root on pressing backspace  1`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-owns="downshift-36-menu"
    data-type="expression-input-root"
    role="combobox"
  >
    <input
      aria-activedescendant="downshift-36-item-0"
      aria-autocomplete="list"
      aria-controls="downshift-36-menu"
      aria-labelledby="downshift-36-label"
      autocomplete="off"
      data-valid="true"
      data-value-type="string"
      id="downshift-35-input"
      style="width: 8px;"
      value=""
    />
    <div
      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
    />
    <ul
      aria-labelledby="downshift-36-label"
      data-type="expression-list"
      id="downshift-36-menu"
      role="listbox"
      style="background-color: white; font-weight: normal; position: absolute;"
    >
      <li
        aria-selected="true"
        data-item-type="function"
        data-type="expression-list-item"
        id="downshift-36-item-0"
        role="option"
        style="background-color: rgb(242, 242, 242);"
      >
        <span>
          SPLIT
        </span>
      </li>
      <li
        aria-selected="false"
        data-item-type="function"
        data-type="expression-list-item"
        id="downshift-36-item-1"
        role="option"
      >
        <span>
          CONCAT
        </span>
      </li>
      <li
        aria-selected="false"
        data-item-type="function"
        data-type="expression-list-item"
        id="downshift-36-item-2"
        role="option"
      >
        <span>
          SUB
        </span>
      </li>
      <li
        aria-selected="false"
        data-item-type="function"
        data-type="expression-list-item"
        id="downshift-36-item-3"
        role="option"
      >
        <span>
          EXTRACT
        </span>
      </li>
      <li
        aria-selected="false"
        data-item-type="dimension"
        data-type="expression-list-item"
        id="downshift-36-item-4"
        role="option"
      >
        <span>
          Account
        </span>
      </li>
      <li
        aria-selected="false"
        data-item-type="dimension"
        data-type="expression-list-item"
        id="downshift-36-item-5"
        role="option"
      >
        <span>
          Ad
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Keyboard events work properly Should navigate to the expression root on pressing left on first param 1`] = `
<DocumentFragment>
  <span
    class=""
    data-type="expression-root"
    style="display: flex;"
    tabindex="0"
  >
    CONCAT(
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="downshift-26-menu"
      class=""
      data-type="expression-input-root"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="downshift-26-menu"
        aria-labelledby="downshift-26-label"
        autocomplete="off"
        data-valid="true"
        data-value-type="string"
        id="downshift-25-input"
        placeholder="dim1"
        style="width: 8px;"
        value=""
      />
      <div
        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
      >
        dim1
      </div>
    </div>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="downshift-28-menu"
      class=""
      data-type="expression-input-root"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="downshift-28-menu"
        aria-labelledby="downshift-28-label"
        autocomplete="off"
        data-valid="true"
        data-value-type="string"
        id="downshift-27-input"
        placeholder="dim2"
        style="width: 8px;"
        value=""
      />
      <div
        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
      >
        dim2
      </div>
    </div>
    )
  </span>
</DocumentFragment>
`;

exports[`Keyboard events work properly Should navigate to the next param on pressing right  1`] = `
<DocumentFragment>
  <span
    class=""
    data-type="expression-root"
    style="display: flex;"
    tabindex="0"
  >
    CONCAT(
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="downshift-10-menu"
      class=""
      data-type="expression-input-root"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="downshift-10-menu"
        aria-labelledby="downshift-10-label"
        autocomplete="off"
        data-valid="true"
        data-value-type="string"
        id="downshift-9-input"
        placeholder="dim1"
        style="width: 8px;"
        value=""
      />
      <div
        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
      >
        dim1
      </div>
    </div>
    <div
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-owns="downshift-12-menu"
      class=""
      data-type="expression-input-root"
      role="combobox"
    >
      <input
        aria-activedescendant="downshift-12-item-0"
        aria-autocomplete="list"
        aria-controls="downshift-12-menu"
        aria-labelledby="downshift-12-label"
        autocomplete="off"
        data-valid="true"
        data-value-type="string"
        id="downshift-11-input"
        placeholder="dim2"
        style="width: 8px;"
        value=""
      />
      <div
        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre;"
      >
        dim2
      </div>
      <ul
        aria-labelledby="downshift-12-label"
        data-type="expression-list"
        id="downshift-12-menu"
        role="listbox"
        style="background-color: white; font-weight: normal; position: absolute;"
      >
        <li
          aria-selected="true"
          data-item-type="function"
          data-type="expression-list-item"
          id="downshift-12-item-0"
          role="option"
          style="background-color: rgb(242, 242, 242);"
        >
          <span>
            SPLIT
          </span>
        </li>
        <li
          aria-selected="false"
          data-item-type="function"
          data-type="expression-list-item"
          id="downshift-12-item-1"
          role="option"
        >
          <span>
            CONCAT
          </span>
        </li>
        <li
          aria-selected="false"
          data-item-type="function"
          data-type="expression-list-item"
          id="downshift-12-item-2"
          role="option"
        >
          <span>
            SUB
          </span>
        </li>
        <li
          aria-selected="false"
          data-item-type="function"
          data-type="expression-list-item"
          id="downshift-12-item-3"
          role="option"
        >
          <span>
            EXTRACT
          </span>
        </li>
        <li
          aria-selected="false"
          data-item-type="dimension"
          data-type="expression-list-item"
          id="downshift-12-item-4"
          role="option"
        >
          <span>
            Account
          </span>
        </li>
        <li
          aria-selected="false"
          data-item-type="dimension"
          data-type="expression-list-item"
          id="downshift-12-item-5"
          role="option"
        >
          <span>
            Ad
          </span>
        </li>
      </ul>
    </div>
    )
  </span>
</DocumentFragment>
`;
